File: frontend/src/types/graph.ts
/**
 * TypeScript interfaces matching the backend Graph Explorer API schemas.
 */

export interface GraphNode {
  id: string;
  label: string;
  name: string;
  entity_type: EntityType;
  fully_qualified_name: string;
  language: string;
  file_path: string;
  start_line: number | null;
  end_line: number | null;
  visibility: string;
  metadata: Record<string, unknown>;
}

export interface GraphEdge {
  id: string;
  source: string;
  target: string;
  relationship_type: RelationshipType;
  confidence: number;
  source_file: string;
  line_number: number;
}

export interface GraphData {
  nodes: GraphNode[];
  edges: GraphEdge[];
  total_nodes: number;
  total_edges: number;
  has_more: boolean;
}

export interface NeighborData {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export interface SearchResultItem {
  entity: GraphNode;
  match_field: string;
}

export interface SearchResults {
  query: string;
  results: SearchResultItem[];
  total: number;
}

export interface FileInfo {
  id: string;
  relative_path: string;
  filename: string;
  language: string;
  size_bytes: number;
}

export interface RelationshipDetail {
  relationship_type: string;
  source?: GraphNode;
  target?: GraphNode;
}

export interface EntityDetail {
  entity: GraphNode;
  file: FileInfo | null;
  outgoing_relationships: RelationshipDetail[];
  incoming_relationships: RelationshipDetail[];
  outgoing_count: number;
  incoming_count: number;
}

export interface HierarchyFile {
  file: GraphNode;
  entities: GraphNode[];
  entity_count: number;
}

export interface HierarchyData {
  repository: { id: string; name: string; url: string } | null;
  children: HierarchyFile[];
  total_files: number;
  total_entities: number;
}

export interface PathData {
  path_found: boolean;
  nodes: GraphNode[];
  edges: GraphEdge[];
  length: number;
}

export interface ComplexityMetrics {
  files: number;
  entities: number;
  relationships: number;
  types_used: number;
  relationship_types_used: number;
}

export interface GraphStatistics {
  repository_id: string;
  total_nodes: number;
  total_files: number;
  total_entities: number;
  total_edges: number;
  nodes_by_type: Record<string, number>;
  edges_by_type: Record<string, number>;
  density: number;
  avg_degree: number;
  languages: Record<string, number>;
  complexity: ComplexityMetrics;
}

export interface RepositorySummary {
  id: string;
  name: string;
  url: string;
  owner: string | null;
  clone_status: string;
  parser_status: string;
  graph_status: string;
  total_files: number | null;
  total_nodes: number;
  total_edges: number;
  nodes_by_label: Record<string, number>;
  edges_by_type: Record<string, number>;
}

// Enums
export type EntityType =
  | 'class'
  | 'interface'
  | 'function'
  | 'method'
  | 'import'
  | 'module'
  | 'file'
  | 'repository'
  | 'external'
  | 'struct'
  | 'namespace'
  | 'constructor'
  | 'unknown';

export type RelationshipType =
  | 'IMPORTS'
  | 'CALLS'
  | 'EXTENDS'
  | 'IMPLEMENTS'
  | 'USES'
  | 'DEPENDS_ON'
  | 'REFERENCES'
  | 'CONTAINS'
  | 'DEFINES'
  | 'DEFINED_IN'
  | 'BELONGS_TO';

export const NODE_TYPE_OPTIONS: EntityType[] = [
  'file', 'class', 'interface', 'function', 'method', 'import', 'module', 'external',
];

export const RELATIONSHIP_TYPE_OPTIONS: RelationshipType[] = [
  'IMPORTS', 'CALLS', 'EXTENDS', 'IMPLEMENTS', 'USES', 'DEPENDS_ON',
  'REFERENCES', 'CONTAINS', 'DEFINES',
];

// View state types
export type ViewMode = 'dashboard' | 'explorer' | 'analytics';

export interface FilterState {
  nodeTypes: EntityType[];
  relationshipTypes: RelationshipType[];
}

File: backend/app/schemas/graph.py
"""
Pydantic response schemas for the Graph Explorer API.
These models define the contract between backend and frontend for all
graph visualization data.
"""
import uuid
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, ConfigDict


class GraphNode(BaseModel):
    """A single node in the visualization graph."""
    id: str
    label: str
    name: str
    entity_type: str
    fully_qualified_name: str = ""
    language: str = ""
    file_path: str = ""
    start_line: Optional[int] = None
    end_line: Optional[int] = None
    visibility: str = ""
    metadata: Dict[str, Any] = {}


class GraphEdge(BaseModel):
    """A single edge (relationship) in the visualization graph."""
    id: str
    source: str
    target: str
    relationship_type: str
    confidence: float = 1.0
    source_file: str = ""
    line_number: int = 0


class GraphData(BaseModel):
    """Paginated graph data containing nodes and edges."""
    nodes: List[GraphNode] = []
    edges: List[GraphEdge] = []
    total_nodes: int = 0
    total_edges: int = 0
    has_more: bool = False


class NeighborData(BaseModel):
    """Neighborhood expansion result for a single node."""
    nodes: List[GraphNode] = []
    edges: List[GraphEdge] = []


class SearchResultItem(BaseModel):
    """A single search result with the matched entity and match info."""
    entity: GraphNode
    match_field: str = "name_contains"


class SearchResults(BaseModel):
    """Collection of search results."""
    query: str
    results: List[SearchResultItem] = []
    total: int = 0


class RelationshipDetail(BaseModel):
    """A relationship with its connected entity detail."""
    relationship_type: str
    source: Optional[GraphNode] = None
    target: Optional[GraphNode] = None


class FileInfo(BaseModel):
    """File context for an entity."""
    id: str
    relative_path: str = ""
    filename: str = ""
    language: str = ""
    size_bytes: int = 0


class EntityDetail(BaseModel):
    """Full detail for a single entity including relationships."""
    entity: GraphNode
    file: Optional[FileInfo] = None
    outgoing_relationships: List[RelationshipDetail] = []
    incoming_relationships: List[RelationshipDetail] = []
    outgoing_count: int = 0
    incoming_count: int = 0


class HierarchyFile(BaseModel):
    """A file in the hierarchy tree with its child entities."""
    file: GraphNode
    entities: List[GraphNode] = []
    entity_count: int = 0


class RepositoryInfo(BaseModel):
    """Minimal repository info for hierarchy."""
    id: str
    name: str
    url: str = ""


class HierarchyData(BaseModel):
    """Tree-structured hierarchy: Repository -> Files -> Entities."""
    repository: Optional[RepositoryInfo] = None
    children: List[HierarchyFile] = []
    total_files: int = 0
    total_entities: int = 0


class PathData(BaseModel):
    """Shortest path result between two entities."""
    path_found: bool = False
    nodes: List[GraphNode] = []
    edges: List[GraphEdge] = []
    length: int = 0


class ComplexityMetrics(BaseModel):
    """Complexity metrics for the graph."""
    files: int = 0
    entities: int = 0
    relationships: int = 0
    types_used: int = 0
    relationship_types_used: int = 0


class GraphStatistics(BaseModel):
    """Comprehensive graph analytics for a repository."""
    repository_id: str
    total_nodes: int = 0
    total_files: int = 0
    total_entities: int = 0
    total_edges: int = 0
    nodes_by_type: Dict[str, int] = {}
    edges_by_type: Dict[str, int] = {}
    density: float = 0.0
    avg_degree: float = 0.0
    languages: Dict[str, int] = {}
    complexity: ComplexityMetrics = ComplexityMetrics()


class RepositorySummary(BaseModel):
    """Repository with graph status and summary counts for the dashboard."""
    id: uuid.UUID
    name: str
    url: str
    owner: Optional[str] = None
    clone_status: str = "pending"
    parser_status: str = "pending"
    graph_status: str = "pending"
    total_files: Optional[int] = None
    total_nodes: int = 0
    total_edges: int = 0
    nodes_by_label: Dict[str, int] = {}
    edges_by_type: Dict[str, int] = {}

    model_config = ConfigDict(from_attributes=True)
